File: CP/UIU Long/Binary Search/binary_search.cpp
#include <iostream>
#include <vector>
using namespace std;

int binary_search(vector<int>& a, int key){
    int low = 0;
    int high = a.size() - 1;

    while(low <= high) {
        int mid = (low + high) / 2;

        if(a[mid] == key) {
            return mid;
        }
        else if(a[mid] > key) {
            high = mid - 1;
        }
        else {
            low = mid + 1;
        }
    }
}

int main(){
    vector<int> a{10, 20, 30, 40, 50, 60, 70};
    int key = 10;
    cout << "Key found at index: " << binary_search(a, key) << "\n";

    return 0;
}

File: Data Structures and Algorithms/4. Recursion/Kunal/Intro/BinarySearch.java
import java.util.*;
public class BinarySearch {

    public static int search(int[] arr, int target, int start, int end) {
        if(start > end) return -1;

        int mid = (start + end) / 2;
        if(arr[mid] == target) {
            return mid;
        } else if(target < arr[mid]) {
            return search(arr, target, start, mid - 1);
        } else {
            return search(arr, target, mid + 1, end);
        }
    }

    public static void print_array(int []arr) {
        int n = arr.length;
        for(int i = 0; i < n; i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int []arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = sc.nextInt();
        }
        print_array(arr);
        int target = sc.nextInt();
        int result = search(arr, target, 0, arr.length - 1);
        System.out.printf("Target found at index %d\n", result);
    }
}
